src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/servlet/CompleteFilterAnnotationQuickFix.java

```diff
@@ -66,65 +78,83 @@ protected void insertAnnotations(Diagnostic diagnostic, JavaCodeActionContext co
-        // API
-
-
-        // if missing an attribute, do value insertion
+    @Override
+    public CodeAction resolveCodeAction(JavaCodeActionResolveContext context) {
+        final CodeAction toResolve = context.getUnresolved();
         PsiElement node = null;
         PsiModifierListOwner parentType = null;
         PsiAnnotation annotationNode = null;
-        if (diagnostic.getCode().getLeft().equals(ServletConstants.DIAGNOSTIC_CODE_FILTER_MISSING_ATTRIBUTE)) {
-            ArrayList<String> attributes = new ArrayList<>();
-            attributes.add("value"); attributes.add("urlPatterns");attributes.add("servletNames");
-            // Code Action 1: add value attribute to the WebServlet annotation
-            // Code Action 2: add urlPatterns attribute to the WebServlet annotation
-            for (int i = 0; i < attributes.size(); i++) {
-                String attribute = attributes.get(i);
-                JavaCodeActionContext targetContext = context.copy();
-                node = targetContext.getCoveringNode();
-                parentType = getBinding(node);
-                annotationNode = getAnnotation(node);
-
-                ArrayList<String> attributesToAdd = new ArrayList<>();
-                attributesToAdd.add(attribute);
+        CodeActionResolveData data = (CodeActionResolveData) toResolve.getData();
+        String diagnosticCode = (String) data.getExtendedDataEntry(DIAGNOSTIC_CODE_KEY);
+        String attribute = (String) data.getExtendedDataEntry(ATTRIBUTE_KEY);
+        String annotation = (String) data.getExtendedDataEntry(ANNOTATION_KEY);
+
+        if (diagnosticCode.equals(ServletConstants.DIAGNOSTIC_CODE_FILTER_MISSING_ATTRIBUTE)) {
+            node = context.getCoveringNode();
+            parentType = getBinding(node);
+            annotationNode = getAnnotation(node);
+
+            ArrayList<String> attributesToAdd = new ArrayList<>();
+            attributesToAdd.add(attribute);
+            String name = getLabel(annotation, attribute, "Add");
+            ChangeCorrectionProposal proposal = new ModifyAnnotationProposal(name, context.getSource().getCompilationUnit(),
+                    context.getASTRoot(), parentType, annotationNode, 0, annotation, attributesToAdd);
+            try {
+                WorkspaceEdit we = context.convertToWorkspaceEdit(proposal);
+                toResolve.setEdit(we);
+            } catch (Exception e) {
+                LOGGER.log(Level.WARNING, "Unable to create workspace edit for code action.", e);
+            }
+        }
+
+        if (diagnosticCode.equals(ServletConstants.DIAGNOSTIC_CODE_FILTER_DUPLICATE_ATTRIBUTES)) {
+            node = context.getCoveringNode();
+            parentType = getBinding(node);
+            annotationNode = getAnnotation(node);
+
+            ArrayList<String> attributesToRemove = new ArrayList<>();
+            attributesToRemove.add(attribute);
+            String name = getLabel(annotation, attribute, "Remove");
```

Comment:
Should retrieve the title from the code action instead of recomputing it here.

Reply:
Made the necessary changes so that the title is retrieved from `toResolve.getTitle()`.